include/range/v3/detail/bind_back.hpp

```diff
@@ -0,0 +1,113 @@
+
+    namespace detail
+    {
+        template<int args_count, typename Fn, typename... Args>
+        class bind_back_t
+        {
+            using Self = bind_back_t<args_count, Fn, Args...>;
+
+            // clang-format off
+            template<typename Self, std::size_t... I, typename... CallArgs>
+            constexpr auto static CPP_auto_fun(apply)
+                (Self & self, std::index_sequence<I...>, CallArgs &&... call_args)
+            (
+                return self.fn(std::forward<CallArgs>(call_args)...,
+                              std::get<I>(self.args_tuple)...)
+            )
+            template<typename Self, typename... CallArgs>
+            constexpr auto static CPP_auto_fun(run)(Self & self, CallArgs &&... call_args)
+            (
+                return apply(self,
+                             std::index_sequence_for<Args...>{},
+                             std::forward<CallArgs>(call_args)...)
+            )
+
+            template<int i>
+            constexpr void static check_args_count() noexcept {
+                static_assert(args_count < 0 || args_count == i,
+                    "Wrong arguments count passed to binded function");
+            }
+
+            Fn fn;
+            std::tuple<Args...> args_tuple;
+            // clang-format on
+
+        public:
+            constexpr bind_back_t(Fn && fn, Args &&... args)
+              : fn(std::move(fn))
+              , args_tuple(std::move(args)...)
+            {}
+
+            template<typename... CallArgs>
+            constexpr decltype(auto) operator()(CallArgs &&... call_args) noexcept(
+                noexcept(run(std::declval<Self &>(),
+                             std::forward<CallArgs>(call_args)...)))
+            {
+                check_args_count<sizeof...(CallArgs)>();
+                return Self::run(*this, std::forward<CallArgs>(call_args)...);
+            }
+            template<typename... CallArgs>
+            constexpr decltype(auto) operator()(CallArgs &&... call_args) const
+                noexcept(noexcept(run(std::declval<const Self &>(),
+                                      std::forward<CallArgs>(call_args)...)))
+            {
+                check_args_count<sizeof...(CallArgs)>();
+                return Self::run(*this, std::forward<CallArgs>(call_args)...);
+            }
+        };
+
+        // T/T&& => T&&, T& => T
+        template<typename T,
+                 typename U = meta::if_<             //
+                     std::is_lvalue_reference<T>,    //
+                     uncvref_t<T>,                   //
+                     std::remove_reference_t<T> &&>> //
+        constexpr U bind_back_forward(std::remove_reference_t<T> & t) noexcept
+        {
+            return static_cast<U>(t);
+        }
+    } // namespace detail
+
+    // clang-format off
+    // args_count >= 0 constraint arguments count on call
+    template<int args_count = -1, typename Fn, typename... Args>
```

Comment:
I am not partial to the `arg_count` template parameter. My weak preference is to drop it as needless complication.

Reply:
I disagree. It provides additional layer of compile time safety. Without it, you mistakenly may pass additional argument(s), which may result in call of "wrong" `operator()()` overload.

For example:
`bind_back<1>(drop, n)` will result in `fn(auto&&)`, while `bind_back(drop, n)` => `fn(auto&&...)`